File: isma-intg-lib/isma-intg-lib-rk2/src/main/java/ru/nstu/isma/intg/lib/rungeKutta/rk22/internal/Rk2StabilityIntgController.java
package ru.nstu.isma.intg.lib.rungeKutta.rk22.internal;

import ru.nstu.isma.intg.api.calcmodel.DaeSystem;
import ru.nstu.isma.intg.api.methods.IntgPoint;
import ru.nstu.isma.intg.core.methods.BaseStabilityIntgController;
import ru.nstu.isma.intg.core.methods.utils.MathUtils;

/**
 * @author Mariya Nasyrova
 * @since 29.08.14
 */
public class Rk2StabilityIntgController extends BaseStabilityIntgController {

    private static final double STABILITY_INTERVAL = 2.0;

    @Override public double getStabilityInterval() {
        return STABILITY_INTERVAL;
    }

    @Override public double getMaxJacobiEigenValue(IntgPoint point) {
        int varCount = point.getStages().length;
        double[] deltaKs = new double[varCount];
        double k1;
        double k2;

        for (int i = 0; i < varCount; i++) {
            k1 = point.getStages()[i][0];
            k2 = point.getStages()[i][1];

            if (k2 == k1) {
                deltaKs[i] = 0.0;
            } else {
                deltaKs[i] = Math.abs(point.getRhs()[DaeSystem.RHS_DE_PART_IDX][i] - k2) / Math.abs(k2 - k1);
            }
        }

        return  2.0 * MathUtils.max(deltaKs);
    }
}

File: isma-intg-lib/isma-intg-lib-rk2/src/main/kotlin/ru/nstu/isma/intg/lib/rungeKutta/rk22/internal/Rk2StabilityIntgController.kt
package ru.nstu.isma.intg.lib.rungeKutta.rk22.internal

import ru.nstu.isma.intg.api.calcmodel.DaeSystem
import ru.nstu.isma.intg.api.methods.IntgPoint
import ru.nstu.isma.intg.core.methods.BaseStabilityIntgController
import ru.nstu.isma.intg.core.methods.utils.MathUtils
import kotlin.math.abs

/**
 * @author Mariya Nasyrova
 * @since 29.08.14
 */
class Rk2StabilityIntgController : BaseStabilityIntgController() {
    public override fun getStabilityInterval(): Double {
        return STABILITY_INTERVAL
    }

    public override fun getMaxJacobiEigenValue(point: IntgPoint): Double {
        val varCount = point.stages.size
        val deltaKs = DoubleArray(varCount)
        var k1: Double
        var k2: Double

        for (i in 0 until varCount) {
            k1 = point.stages[i][0]
            k2 = point.stages[i][1]

            if (k2 == k1) {
                deltaKs[i] = 0.0
            } else {
                deltaKs[i] = abs(point.rhs[DaeSystem.RHS_DE_PART_IDX][i] - k2) / abs(k2 - k1)
            }
        }

        return 2.0 * MathUtils.max(deltaKs)
    }

    companion object {
        private const val STABILITY_INTERVAL = 2.0
    }
}
